utils/protobuf-schema-utilities/src/main/java/io/apicurio/registry/utils/protobuf/schema/FileDescriptorUtils.java

```diff
@@ -191,6 +210,152 @@ public static FileDescriptor protoFileToFileDescriptor(String schemaDefinition,
+            protoFileElements.put(protoFullName, protoFile);
+            schemaContents.put(protoFullName, schemaContent);
+        }
+    }
+
+    public static FileDescriptor parseProtoFileWithDependencies(String schemaDefinition, String protoFileName,
+                                                                Map<String, String> requiredSchemaDependencies)
+            throws DescriptorValidationException {
+        Objects.requireNonNull(schemaDefinition);
+        Objects.requireNonNull(protoFileName);
+        Objects.requireNonNull(requiredSchemaDependencies);
+
+        final Map<String, FileDescriptor> resolvedDependencies = mutableBaseDependenciesByName(requiredSchemaDependencies.size());
+        final ProtoFileElement mainProtoElement = ProtoParser.Companion.parse(DEFAULT_LOCATION, schemaDefinition);
+        return resolveFileDescriptor(mainProtoElement, schemaDefinition, protoFileName, requiredSchemaDependencies,
+                resolvedDependencies, null, new HashSet<>(), null);
+    }
+
+    private static FileDescriptor resolveFileDescriptor(ProtoFileElement mainProtoElement,
+                                                        String schemaDefinition,
+                                                        String protoFileName,
+                                                        Map<String, String> schemaDefinitions,
+                                                        Map<String, FileDescriptor> resolvedDependencies,
+                                                        Map<String, String> requiredDependentSchemas,
+                                                        Set<String> unresolvedImportNames,
+                                                        Map<String, ProtoFileElement> cachedProtoFileDependencies) throws DescriptorValidationException {
+        final String mainProtoImportName = toProtoFullName(mainProtoElement, protoFileName);
+        if (!unresolvedImportNames.add(mainProtoImportName)) {
+            // TODO we can do better here, we can actually print the whole chain of dependencies
+            throw new IllegalStateException("Circular Dependency found");
+        }
+        List<String> directDependencyNames = mainProtoElement.getImports();
+        if (requiredDependentSchemas == null) {
+            requiredDependentSchemas = new HashMap<>(directDependencyNames.size());
+        }
+        // TODO we can make a singleton of empty fd
+        final FileDescriptor[] directDependencyFds = new FileDescriptor[directDependencyNames.size()];
+        for (int i = 0; i < directDependencyFds.length; i++) {
+            final String depFullName = directDependencyNames.get(i);
+            FileDescriptor fdDep = resolvedDependencies.get(depFullName);
+            final String schemaDep = schemaDefinitions.get(depFullName);
+            // this has never been resolved before
+            if (fdDep == null) {
+                final String fileName = extractProtoFileName(depFullName);
```

Comment:
Here, by checking that a schema's dep is null we can fast-fail and report back we have a missing required dependency: it can both help to report the current main proto we cannot resolve, due to missing deps and, by using unresolvedImportNames (excluding the current main proto), we can state which others protos won't be resolved, too.